java/http/vertx/java-http-vertx-consumer/src/main/java/HttpKafkaConsumer.java

```diff
@@ -68,190 +69,192 @@ public void start(Future<Void> startFuture) throws Exception {
-        .compose(v -> {
-            this.pollTimer = vertx.setPeriodic(this.config.getPollInterval(), t -> {
-                this.poll().setHandler(ar -> {
-                    if (ar.succeeded()) {
-                        log.info("Received {}", ar.result());
-                    }
-                });
-            });
-            startFuture.complete();
-        }, startFuture);
+                .compose(consumer -> this.subscribe(consumer, this.config.getTopic()))
+                .compose(v -> {
+                    this.pollTimer = vertx.setPeriodic(this.config.getPollInterval(), t -> {
+                        this.poll().onComplete(ar -> {
+                            log.info("Received {} ", ar);
+                        }).onFailure(cause -> {
+                            log.error("Failed to poll", cause);
+                        });
+                    });
+                    startPromise.complete();
+                    return Future.succeededFuture();
+                })
+                .onFailure(startPromise::fail);
     }
 
     @Override
-    public void stop(Future<Void> stopFuture) throws Exception {
+    public void stop(Promise<Void> stopPromise) throws Exception {
         log.info("HTTP Kafka consumer stopping");
         if (this.consumer != null) {
             this.vertx.cancelTimer(this.pollTimer);
-            this.deleteConsumer().setHandler(ar -> {
-                stopFuture.complete();
+            this.deleteConsumer().future().onComplete(ar -> {
+                stopPromise.complete();
             });
         } else {
-            stopFuture.complete();
+            stopPromise.complete();
         }
     }
 
+    private Future<Void> subscribe(CreatedConsumer consumer, String topic) {
+        Promise<Void> promise = Promise.promise();
+
+        JsonObject topics = new JsonObject()
+                .put("topics", new JsonArray().add(topic));
+
+        HttpRequest<JsonObject> request = client.post(consumer.getBaseUri() + "/subscription")
+                .putHeader(HttpHeaderNames.CONTENT_LENGTH.toString(), String.valueOf(topics.toBuffer().length()))
+                .putHeader(HttpHeaderNames.CONTENT_TYPE.toString(), "application/vnd.kafka.v2+json")
+                .as(BodyCodec.jsonObject());
+
+        request.sendJsonObject(topics).onComplete(ar -> {
+            if (ar != null && ar.succeeded()) {
+                HttpResponse<JsonObject> response = ar.result();
+                if (response.statusCode() == HttpResponseStatus.NO_CONTENT.code()) {
+                    log.info("Subscribed to {}", topic);
+                    promise.complete();
+                } else {
+                    promise.fail(new RuntimeException("Got HTTP status code " + response.statusCode()));
+                }
+            } else {
+                promise.fail(ar != null ? ar.cause() : new NullPointerException("ar is null"));
+            }
+        });
+        return promise.future();
```

Comment:
Just a tip: if you leave this method where the original was, the diff will look nicer and reviews will be much easier for reviewers as they would see the old and new code next to each other. There does not seem to be any real reason why to move this method within the same class.

Reply:
Not sure why I moved the subscribe method. Put back to correct place.